src/actions/tamanu/psql.rs

```diff
@@ -0,0 +1,112 @@
+use std::fs;
+
+use clap::Parser;
+use miette::{miette, Context as _, IntoDiagnostic, Result};
+use tracing::{debug, instrument};
+
+use crate::actions::Context;
+
+use super::config::{merge_json, package_config};
+use super::{find_tamanu, TamanuArgs};
+
+/// Connect to Tamanu's db via `psql`.
+#[derive(Debug, Clone, Parser)]
+pub struct PsqlArgs {
+	/// Package to look at
+	#[arg(short, long)]
+	pub package: String,
+
+	/// Include defaults
+	#[arg(short = 'D', long)]
+	pub defaults: bool,
+
+	/// If given, this overwrites the username in the config and prompts for a password
+	#[arg(short, long)]
+	pub username: Option<String>,
+}
+
+/// The Tamanu config only describing the part `psql` needs
+#[derive(serde::Deserialize, Debug)]
+struct Config {
+	db: Db,
+}
+
+#[derive(serde::Deserialize, Debug)]
+struct Db {
+	name: String,
+	username: String,
+	password: String,
+}
+
+pub async fn run(ctx: Context<TamanuArgs, PsqlArgs>) -> Result<()> {
+	let (_, root) = find_tamanu(&ctx.args_top)?;
+
+	let config_value = if ctx.args_sub.defaults {
+		merge_json(
+			package_config(&root, &ctx.args_sub.package, "default.json5")?,
+			package_config(&root, &ctx.args_sub.package, "local.json5")?,
+		)
+	} else {
+		package_config(&root, &ctx.args_sub.package, "local.json5")?
+	};
+
+	let config: Config = serde_json::from_value(config_value)
+		.into_diagnostic()
+		.wrap_err("parsing of Tamanu config failed")?;
+	let name = config.db.name;
+	let (username, password) = if let Some(ref username) = ctx.args_sub.username {
+		// Rely on `psql` password prompt by making the password parameter empty.
+		(username.as_str(), "")
+	} else {
+		(config.db.username.as_str(), config.db.password.as_str())
+	};
+
+	// By default, consoles on Windows use a different codepage from other parts of the system.
+	// What that implies for us is not clear, but this code is here just in case.
+	// See https://www.postgresql.org/docs/current/app-psql.html
+	#[cfg(windows)]
+	unsafe {
+		windows::Win32::System::Console::SetConsoleCP(1252).into_diagnostic()?
+	}
+
+	let psql_path = find_psql().wrap_err("failed to find psql executable")?;
+	// Use the default host, which is the localhost via Unix-domain socket on Unix or TCP/IP on Windows
+	duct::cmd!(psql_path, "--dbname", name, "--username", username,)
+		.env("PGPASSWORD", password)
+		.env("PSQL_HISTORY", root.with_file_name("psql.history"))
+		.run()
+		.into_diagnostic()
+		.wrap_err("failed to execute psql")?;
+
+	Ok(())
+}
+
+#[instrument(level = "debug")]
+fn find_psql() -> Result<String> {
```

Comment:
For correctness this should probably return `PathBuf`. Then you can construct the path with `.join` instead of `format!`.